packages/react-composition/src/Context.tsx

```diff
@@ -0,0 +1,139 @@
+import React, {
+    FC,
+    ComponentType,
+    useState,
+    useCallback,
+    createContext,
+    useContext,
+    useMemo
+} from "react";
+
+export function compose(...fns: HigherOrderComponent[]) {
+    return function ComposedComponent(Base: FC<unknown>): FC<unknown> {
+        return fns.reduceRight((Component, hoc) => hoc(Component), Base);
+    };
+}
+
+interface ComposedComponent {
+    /**
+     * Ready to use React component.
+     */
+    component: ComponentType<unknown>;
+    /**
+     * HOCs used to compose the original component.
+     */
+    hocs: HigherOrderComponent[];
+}
+
+/**
+ * IMPORTANT: TInputProps default type is `any` because this interface is use as a prop type in the `Compose` component.
+ * You can pass any HigherOrderComponent as a prop, regardless of its TInputProps type. The only way to allow that is
+ * to let it be `any` in this interface.
+ */
+export interface HigherOrderComponent<TInputProps = any, TOutputProps = TInputProps> {
+    (Component: FC<TInputProps>): FC<TOutputProps>;
+}
+
+type ComposedComponents = Map<ComponentType<unknown>, ComposedComponent>;
+
+interface CompositionContext {
+    components: ComposedComponents;
+    getComponent(component: ComponentType<unknown>): ComponentType<unknown> | undefined;
+    composeComponent(component: ComponentType<unknown>, hocs: HigherOrderComponent[]): void;
+}
```

Comment:
Quite a lot of typescript interfaces & types here and above, worth abstracting into their own file?

Reply:
I like to keep internal types colocated with code, especially because they're not exported for public usage. It's also easier to maintain them this way. Not all types need to be exported from the package root, so I kept these here.